wallet/rpc: sendrawtransaction maxfeerate

kallewoof opened the pull request:
This adds a new `maxfeerate` parameter to `sendrawtransaction` which forces the node to reject a transaction whose feerate is above the given fee rate.

This is a safety harness from shooting yourself in the foot and accidentally overpaying fees.

See also #12911.

kallewoof committed:
wallet/rpc: add maxfeerate parameter to sendrawtransaction

NicolasDorier commented:
Your PR show me actually that `maxtxfee` exists, which is uncorrelated with absurd fee we were talking about.

One could say that changing `maxtxfee` or having `maxtxfeerate` would solve the issue without touching RPC. But that said, when you do a transaction manually, you know the `maxfeerate` to expect when you call `sendrawtransaction`, not when you modify the config file of bitcoind.

ACK.

(That said, I think this would be better in `signrawtransaction*` than `sendrawtransaction`)

kallewoof commented:
> Your PR show me actually that maxtxfee exists, which is uncorrelated with absurd fee we were talking about.

Yes, absurdfee can actually be specified for each call to `AcceptToMempool`, so it can be used after all.

> One could say that changing `maxtxfee` or having `maxtxfeerate` at config level would solve the issue without touching RPC. But, when you do a transaction manually, you know the `maxfeerate` to expect when you call `sendrawtransaction`, not when you modify the config file of bitcoind.

I think having a configurable default maxtxfee would be a good idea. Perhaps this PR should then ignore the default if the user does `allowhighfees=true`.

> (That said, I think this would be better in `signrawtransaction*` than `sendrawtransaction`)

It fits better here, because it already has the functionality needed in the `AcceptToMemoryPool` function via the absurd fee feature. In `sign`, you sometimes do not know the input amounts, which means the system sometimes can't check. Here, the amounts are always known because the mempool either knows or the tx is invalid.

NicolasDorier commented:
my point is `maxtxfee` already exists. `allowhighfees=true` use it already, it is not absurd fee as I thought.

kallewoof commented:
You mean reuse the argument instead of adding a new one? So boolean for current behavior and BTC for feerate?

promag commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -378,5 +378,34 @@ def run_test(self):
+        addr = self.nodes[2].getnewaddress()
+        txId = self.nodes[0].sendtoaddress(addr, 1.0)
+        rawTx = self.nodes[0].getrawtransaction(txId, True)
+        vout = False
```
Alternative (not tested)
```python
vout = next(o for o in rawTx['vout'] if o['value'] == Decimal('1.00000000'))
```

kallewoof replied:
Several copies of this in `rpc_rawtransaction`, so I made a dedicated commit to address the others.

kallewoof committed:
test: Refactor vout fetches in rpc_rawtransaction

kallewoof commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1123,6 +1122,13 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
+        size_t weight = GetTransactionWeight(*tx);
+        CFeeRate fr(AmountFromValue(request.params[2]));
+        CAmount max_fee = fr.GetFee((weight+3)/4);
+        nMaxRawTxFee = nMaxRawTxFee > 0 ? std::min(nMaxRawTxFee, max_fee) : max_fee;
```
I am sort of overriding `allowhighfees` right now. I could disallow it completely, i.e. require that it is null or `false`, if using `maxfeerate`, but not sure it's worth it.

kallewoof replied:
To more directly answer your question, though, `allowhighfees=true` will simply remove the existing absurd fee cap (which is 0.1 btc I believe), so if maxfeerate was really high, it actually could result in a slightly higher cap.

luke-jr replied:
Just have a single parameter... if it's a boolean, it's the (deprecated?) allowhighfees; if a number, then a specific limit.

promag replied:
@luke-jr does that work with named parameters?

kallewoof replied:
> does that work with named parameters?

Yes, it is already used by `getbalance`:

https://github.com/bitcoin/bitcoin/blob/222e627322ce4de3292259a4868d23983f2a5394/src/wallet/rpcwallet.cpp#L4784

kallewoof replied:
Actually, I'm not sure if it works. I'm having some trouble getting it to work. Maybe if I use the same name for both. (Have to run away for the day but will try tomorrow.)

kallewoof replied:
Can't make separate names work so overloading `allowhighfees` to mean flag when bool, and max fee rate when numeric.

kallewoof replied:
Turns out the issue was elsewhere, so back to using separate names.

MeshCollider commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1084,12 +1084,12 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
             "\nArguments:\n"
             "1. \"hexstring\"    (string, required) The hex string of the raw transaction)\n"
-            "2. allowhighfees    (boolean, optional, default=false) Allow high fees\n"
+            "2. allowhighfees|maxfeerate (boolean/numeric, optional, default=false/∞) Allow high fees (boolean) or reject transactions whose fee rate is higher than the specified value (numeric), expressed in " + CURRENCY_UNIT + "/kB\n"
```
Shouldn't this default be whatever `maxTxFee` is, not `∞`?

kallewoof replied:
Fixed - it now outputs
```
2. allowhighfees|maxfeerate (boolean/numeric, optional, default=false/0.10) Allow 
high fees (boolean) or reject transactions whose fee rate is higher than the 
specified value (numeric), expressed in BTC/kB
```
for the default settings.

MeshCollider commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1115,8 +1113,15 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
+    } else if (request.params[1].isNum()) {
+        size_t weight = GetTransactionWeight(*tx);
+        CFeeRate fr(AmountFromValue(request.params[1]));
+        nMaxRawTxFee = fr.GetFee((weight+3)/4);
```
Might be good to add a quick comment here to clarify the `+3/4` is for rounding

kallewoof replied:
Added note.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/13541/commits/dbd466c212ee22f1cab4dd6c3b684f6efb78ef4a

MarcoFalke commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1021,7 +1022,7 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
                 {
                     {"hexstring", RPCArg::Type::STR_HEX, /* opt */ false, /* default_val */ "", "The hex string of the raw transaction"},
-                    {"allowhighfees", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "false", "Allow high fees"},
+                    {"maxfeerate", RPCArg::Type::AMOUNT, /* opt */ true,  /* default_val */ FormatMoney(maxTxFee), "Allow high fees (boolean) or reject transactions whose fee rate is higher than the specified value (numeric), expressed in " + CURRENCY_UNIT + "/kB\n"},
```
```suggestion
                    {"maxfeerate", RPCArg::Type::AMOUNT, /* default */ FormatMoney(maxTxFee), "Allow high fees (boolean) or reject transactions whose fee rate is higher than the specified value (numeric), expressed in " + CURRENCY_UNIT + "/kB\n"},
```

MarcoFalke commented:
Could do the same for `testmempoolaccept` (since that is just a mirror of sendraw)

kallewoof committed:
wallet/rpc: add maxfeerate parameter to testmempoolaccept

kallewoof commented:
@MarcoFalke Done!

kallewoof commented:
I also tentatively moved `maxTxFee` from validation into wallet in fc29a5a -- I don't know if this should be a separate PR or not, but it seems this move is desired?

jnewbery commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1052,20 +1053,29 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
+        size_t weight = GetTransactionWeight(*tx);
+        CFeeRate fr(AmountFromValue(request.params[1]));
+        // the +3/4 part rounds the value up, and is the same formula used when
+        // calculating the fee for a transaction
```
Can you refer to `GetVirtualTransactionSize()` in this comment? (It took me some time to find and confirm this formula)

kallewoof replied:
Makes sense - added comment to both places.

jnewbery reviewed (commented):
These changes seem reasonable, although I think the commits could be re-arranged and squashed. For example, the first two commits don't cause any behaviour change and the third commit *test: test maxfeerate parameter in sendrawtransaction* tests behaviour that has not been implemented, meaning the test fails. I'd suggest the following:

- One commit for refactoring `BroadcastTransaction()`
- One commit for refactoring `rpc_rawtransaction.py (*test: Refactor vout fetches in rpc_rawtransaction*)
- One commit for changing `sendrawtransaction` together with test updates.
- One commit for changing `testmempoolaccept` together with test updates.

MarcoFalke commented:
@jnewbery You might want to take a look at  #15408, which is a fixup to #14978 (the pull request that moved ::maxtxfee from the rpc callers to the util method)

MarcoFalke commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1035,7 +1036,7 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
                 {
                     {"hexstring", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The hex string of the raw transaction"},
-                    {"allowhighfees", RPCArg::Type::BOOL, /* default */ "false", "Allow high fees"},
+                    {"maxfeerate", RPCArg::Type::AMOUNT, /* default */ FormatMoney(maxTxFee), "Allow high fees (boolean) or reject transactions whose fee rate is higher than the specified value (numeric), expressed in " + CURRENCY_UNIT + "/kB\n"},
```
If bool is no longer mentioned here, the transition should be mentioned in the release notes, though

kallewoof replied:
I'm hesitant to allow some hidden underlying default max fee when using a boolean. If we are removing the option, we should remove it completely, not just from the documentation.

I do realize that this means someone who use '1' as boolean true will accept any fees up to 1 BTC/kB, which is rather high. Maybe removing the boolean style completely is actually the way to go.

MarcoFalke replied:
> Maybe removing the boolean style completely is actually the way to go.

I tend to agree with your idea. An option could be to have a `-deprecatedrpc=sendrawtransaction -deprecatedrpc=testmempoolaccept` to toggle back to the bool behavior.

jnewbery replied:
> An option could be to have a `-deprecatedrpc=sendrawtransaction`

That's one way, although I think it may be overkill. How about catching if a bool is passed, and then returning an error message telling the user to set 'maxreerate' to 0, which maintains exactly the same behaviour. Code:

```
    // TODO: temporary migration code for old clients. Remove in v0.21
    if (request.params[1].isBool()) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Second argument must be numeric (maxfeerate) and no longer supports a boolean. To allow a transaction with high fees, set maxfeerate to 0.");
    } else if (request.params[1].isNum()) {
        [...]
```
Or similar.

kallewoof replied:
I think the throw behavior makes sense, unless we hear complaints. I'll put that in. Thanks for code example.

jnewbery commented:
Thanks @MarcoFalke . I wasn't aware of that PR. I've already re-arranged the commits in the order I suggested here: https://github.com/jnewbery/bitcoin/tree/PR13541.1 . I guess it would be possible to remove the first commit from https://github.com/jnewbery/bitcoin/tree/PR13541.1 and rebase on #15408.

kallewoof commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -286,11 +286,7 @@ def run_test(self):
-            if outpoint['value'] == Decimal('2.20000000'):
-                vout = outpoint
-                break
+        vout = next(o for o in rawTx['vout'] if o['value'] == Decimal('2.20000000'))
```
Props to @promag! :)

promag replied:
Thanks, now I have to understand what I've suggested :trollface:

kallewoof commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -435,5 +427,35 @@ def run_test(self):
         decrawtx = self.nodes[0].decoderawtransaction(rawtx)
         assert_equal(decrawtx['version'], 0x7fffffff)
 
+        #########################################################
```
Sure thing.

kallewoof commented:
I --force rest over @jnewbery's branch. I kind of prefer to have tests separate from code changes, but I'm just gonna leave things as they are. Thanks for the help!

MarcoFalke commented on src/rpc/rawtransaction.cpp:
```diff
@@ -28,6 +28,7 @@
 #include <script/standard.h>
 #include <uint256.h>
 #include <util/bip32.h>
+#include <util/moneystr.h>
```
in commit 9c6a8b52394c9fdbb8c2296a77614b1c0114db49

Why the include? I can't see what it is for

kallewoof replied:
It was in the wrong commit. It's used for `FormatMoney` in RPC helps.

jnewbery commented:
> I kind of prefer to have tests separate from code changes

I have no strong preference in the case of new tests, but obviously changes that need to be made to existing tests when behaviour changes should be atomic with the behaviour change commit to not break git bisects.

Sorry I was a bit overzelous in squashing. Obviously it's your choice and I won't be at all offended if you split out the new tests as part of your next rebase :)

kallewoof commented:
@jnewbery Good point, yeah I meant new tests, not fixes. I should probably be more careful and think about bisect.

kallewoof commented:
It kind of came natural with the PR as I was debugging it, but I added fee related output to testmempoolaccept for the case where fee is related to the rejection (`REJECT_HIGHFEE` and `REJECT_INSUFFICIENTFEE`, specifically). This also included updating AcceptToMempool and friends to allow an outgoing fee.

MarcoFalke commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1160,13 +1160,35 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                fees.pushKV("fee-abs-shortage", ValueFromAmount(max_raw_tx_fee - fee_out));
+                fees.pushKV("fee-rate-shortage", ValueFromAmount(fr_shortage.GetFeePerK()));
+            }
+            result_0.pushKV("fees", fees);
```
in commit 349860df8695ff0ad657a3d6d1502edb46571174

This is undocumented, but I am also not sure if it is a desired feature. Might want to split out as a separate pull request after the first part is merged?

jnewbery replied:
I agree that this is unnecessary in this PR. Happy to review in a follow-up PR.

maflcko reviewed (commented):
utACK 685f2f54866e14f79d56c1220abbd37e23b709d8

kallewoof commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1055,20 +1056,30 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
                 },
             }.ToString());
 
-    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VBOOL});
```
Got it!

kallewoof commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1055,20 +1056,30 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
+        // (see GetVirtualTransactionSize)
+        max_raw_tx_fee = fr.GetFee((weight+3)/4);
+    } else if (!request.params[1].isNull()) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "second argument must be a bool (allowhighfees) or numeric (maxfeerate)");
```
Fixed error message.

kallewoof commented on test/functional/feature_fee_estimation.py:
```diff
@@ -65,7 +65,7 @@ def small_txpuzzle_randfee(from_node, conflist, unconflist, amount, min_fee, fee
     for inp in tx.vin:
         inp.scriptSig = SCRIPT_SIG[inp.prevout.n]
-    txid = from_node.sendrawtransaction(ToHex(tx), True)
+    txid = from_node.sendrawtransaction(ToHex(tx), 1.0)
```
Thanks, fixed!

kallewoof commented on src/rpc/client.cpp:
```diff
@@ -95,8 +95,10 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "signrawtransactionwithkey", 2, "prevtxs" },
     { "signrawtransactionwithwallet", 1, "prevtxs" },
     { "sendrawtransaction", 1, "allowhighfees" },
```
I think it makes sense to keep until we remove the bool check, in case somebody is using named style, or is that unnecessary?

kallewoof commented:
Addressed review comments and removed fee info related commits
	121afc4f5c323776584dd83d93c6da77128b6775
	349860df8695ff0ad657a3d6d1502edb46571174

MarcoFalke commented:
Could you fixup the last commit, please?

MarcoFalke commented:
re-utACK 6c98fd0466 otherwise

kallewoof commented:
@MarcoFalke Squashed.

MarcoFalke commented on test/functional/feature_cltv.py:
```diff
@@ -113,7 +113,7 @@ def run_test(self):
         assert_equal(
             [{'txid': spendtx.hash, 'allowed': False, 'reject-reason': '64: non-mandatory-script-verify-flag (Negative locktime)'}],
-            self.nodes[0].testmempoolaccept(rawtxs=[bytes_to_hex_str(spendtx.serialize())], allowhighfees=True)
+            self.nodes[0].testmempoolaccept(rawtxs=[bytes_to_hex_str(spendtx.serialize())], maxfeerate=0)
```
`bytes_to_hex_str` has been removed. Sorry, needs rebase again :(

kallewoof replied:
NP! Rebased.

MarcoFalke commented:
re-utACK 6c17985e8b

maflcko reviewed (commented):
Just some style nits (could be ignored)

kallewoof commented:
Addressed @MarcoFalke nits.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3130,6 +3130,26 @@ static int GetWitnessCommitmentIndex(const CBlock& block)
+        }
+    }
+    return -1;
+}
```
re-utACK 7abd2e697c0f8e93245e09ac853bae05d0b48bee


Sorry I only noticed this unused method in this hunk after merge :crying_cat_face:.

jnewbery commented:
post-merge ACK 7abd2e697c0f8e93245e09ac853bae05d0b48bee

(except the change to validation.cpp :see_no_evil: )
